### ISSUE TITLE
RetroArch won't run on my RP4 (2gb) It does compile

### ISSUE THREAD

**Issue by namanix** [libretro/RetroArch#9025]
I've been trying to get RetroArch to work on the new Raspberry Pi4 (2gb).
But it just won't start. 
I've compiled it with this configure:
`CFLAGS='-mfpu=neon' ./configure --enable-alsa --enable-udev --enable-floathard --enable-neon --enable-dispmanx --disable-opengl1`

With the Legacy Driver I get: 
```
[INFO] === Build =======================================
[INFO] Capabilities: NEON VFPv3 VFPv4 
[INFO] Built: Jun 26 2019
[INFO] Version: 1.7.7
[INFO] =================================================
[INFO] Environ SET_PIXEL_FORMAT: RGB565.
[INFO] Version of libretro API: 1
[INFO] Compiled against API: 1
[INFO] [Audio]: Set audio input rate to: 30000.00 Hz.
[INFO] [Video]: Video @ 960x720
[INFO] [EGL] Falling back to eglGetDisplay
[INFO] [EGL]: EGL version: 1.4
```
After that it just stops and does nothing

And with the GL Driver I get
`Failed to add service. Already in use?`

It seems like there is a bug somewhere in RetroArch or it's simply not really to be run on the Pi4
Does anyone have any idea how I can fix this?

### Steps to reproduce the bug

1. Download stable 1.7.7
2. Compile with the config mentioned above
3. Try to run it

- RetroArch: 1.7.7
- OS: Raspbian Buster (June 2019)

**Comment by gouchi** [comment]
Is it possible to have the output of 
`glxinfo -B`

Thank you.

**Comment by namanix** [comment]
Sure

```
name of display: :0.0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: VMware, Inc. (0xffffffff)
    Device: llvmpipe (LLVM 6.0, 128 bits) (0xffffffff)
    Version: 19.1.0
    Accelerated: no
    Video memory: 1761MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 3.3
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: llvmpipe (LLVM 6.0, 128 bits)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 19.1.0-devel
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.1 Mesa 19.1.0-devel
OpenGL shading language version string: 1.40
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.0 Mesa 19.1.0-devel
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
```

**Comment by natinusala** [comment]
I think you should try the default driver first as it's supposed to work (instead of the legacy driver)

**Comment by namanix** [comment]
I'm going to try to compile RetroArch tonight on Raspbian Lite. I'll let you know how it goes!

**Comment by namanix** [comment]
After more testing someone on Discord finally had an answer. Compiling RetroArch with --disable-videocore seems to fix the issue of "Failed to add service - Already in use?" 
`./configure --disable-videocore --disable-opengl1`
Built and fixed the issue. But still RetroArch would not run.
It seems like Raspbian is shipped with libraries that are not working. I've got this apt-get from someone on discord: 
`sudo apt-get install build-essential libxkbcommon-dev zlib1g-dev libfreetype6-dev libegl1-mesa-dev libgles2-mesa-dev libgbm-dev libavcodec-dev libsdl2-dev libsdl-image1.2-dev libxml2-dev yasm libavformat-dev libavdevice-dev libswresample-dev libavresample-dev libswscale-dev libv4l-dev libgl*-mesa-dev`

Rebuilt Retroarch and it ran (After installing the retroarch-assets).
I'll ask the person on Discord to comment on this issue to maybe explain why installing those libs fixed the issue.

**Comment by ceb33** [comment]
rpi4 is more standard in the way it works, it doesn't need the broadcom library to use opengl (as on rpi 1/2/3) so installing mesa gl drivers is the important part 
rpi fondation worked with mesa so they use the videocore acceleration internally

--disable-videocore then to make sure it doesn't call broadcom functions (eg bcm_host_init...)
--disable-opengl1 as it' snot supported by rpi4

if you add --enable-opengles and --enable-opengles3 then opengl es will also be available (as the rpi 4 have those as well)

**Comment by natinusala** [comment]
Why aren't those libraries installed by default on Raspbian so that
everything works out of the box? That's what's deceptive part here

**Comment by ceb33** [comment]
maybe because it's teh first version of the OS, release before debian even reelase the official buster
or maybe because they want the OS to be light so not include opengl if users don't need those
etc...

**Comment by namanix** [comment]
Would it be possible to add this to the check this while ./configure is running? 
So that people don't compile with the wrong libraries?
It's just a thought :)

**Comment by gouchi** [comment]
So what was at the end the correct configure options ?

Please close this issue afterwards as it is fixed.

Thank  you.

**Comment by namanix** [comment]
I'll be installing a clean OS tonight on my Rpi4 and recompile with the knowledge I have now to write a good solution to the problem and close the ticket then!

**Comment by ceb33** [comment]
--disable-videocore is the only one really required
--disable-opengl1 is needed to disable openGL 1.1 (but i was able to make a compatible build with it)

rpi4 need gles2-mesa-dev, libgles2-mesa-dev and libegl1-mesa-dev to be able to use --enable-opengles and --enable-opengles3
that's about it afaik

**Comment by natinusala** [comment]
Does the glcore gfx driver work?

**Comment by ceb33** [comment]
i was able to use gl, glcore , and sdl yes

**Comment by seven-lynx** [comment]
The main issue I'm having is that even when I get retroarch to compile and install without errors, the Core Updater doesn't work. When I click on it, the screen goes black for a split second but nothing else happens.

If I run '$ retroarch --verbose', when I click Core Updater the log says

[INFO] [GL]: VSync => on
[INFO] [EGL]: eglSwapInterval(1)

I tried downloading cores manually and extracting them to the cores folders, but they don't register within retroarch.

I'm using a Raspberry Pi 4 4gb. I've reformatted and recompiled several times, and it's always the same result.

My first attempt followed this guide:

https://gist.github.com/AlexMax/32e5d038a66ce57253e740ea75736805

The first configuration I tried that resulted in a usable retroarch was:

$ sudo apt install build-essential libasound2-dev libudev-dev libxkbcommon-dev zlib1g-dev libfreetype6-dev libegl1-mesa-dev libgles2-mesa-dev libgbm-dev libavcodec-dev libsdl2-dev libsdl-image1.2-dev libxml2-dev yasm libavformat-dev libavdevice-dev libswresample-dev libavresample-dev libswscale-dev libv4l-dev libgl*-mesa-dev

$ CFLAGS='-mfpu=neon' ./configure --enable-alsa --enable-udev --enable-floathard --enable-neon --enable-dispmanx --disable-opengl1 --disable-videocore

$ make

$ sudo make install

Core Updater doesn't work, and tweaking settings accomplishes nothing.

Next was:

$ CFLAGS='-mfpu=neon' ./configure --enable-alsa --enable-udev --enable-floathard --enable-neon --enable-dispmanx --disable-opengl1 --disable-videocore --enable opengles --enable-opengles3

Same result.

After that I tried:

CFLAGS='-mfpu=neon -mtune=cortex-a72 -march=armv8-a' ./configure --disable-opengl1 --enable-neon --enable-udev --enable-floathard --enable-opengles3 --enable-opengles --disable-videocore --disable-pulse --disable-oss

Same result.

It might be worth nothing that with the first two builds, dispmanx doesn't show up as an option for the video driver.

After these 2 tries, I reformatted again and repeated both methods, just to make sure I hadn't made an error somewhere. This time I got compiler errors related to dispmanx, so I tried leaving --enable-dispmanx out and got no compiler errors, but still Core Updater wasn't working, nor was manually copying cores over.

After a lot of google-fu, I've come across the method directly above:

"CFLAGS="-mfpu=neon" ./configure --enable-alsa --enable-udev --enable-neon --disable-videocore --disable-opengl1

Then make.

This method requires fkms-v3d to be enabled in the boot config.txt

But I'm not sure what fkms-v3d is or how to install it.

I've been playing with the rpi for years, and my habit on getting a new one is to install retropie (with berryboot) or retroarch without, then kodi, then the rest of my projects. It's driving me nuts that I've had the rpi4 for a few days and still can't figure out how to play my old ps1 games on it.

What am I missing here? I made sure to edit retroarch.cfg to point core_updater_buildbot_url to http://buildbot.libretro.com/nightly/linux/armhf/latest/

Some semi-related noobish questions:

How do I format my post so that I can present the commands I've tried the same way everyone else seems to?

What is fkms-v3d? What is x11? These relate to how Raspbian displays windows?

If I want to change the CLFAGS portion of my install, how should I go about recompiling? Should I just delete the RetroArch-1.7.7 folder and start over at $ tar -zxvf v1.7.7.tar.gz, or is there a better way to remove the previous installation and start over?

Has anyone figured out a berryboot type method of dual booting the rpi4? If so, I'd like to try RetroPie and see if I have the same issue.

Save me, you guys. I'm getting frustrated. Also I apologize if this isn't the correct place to post this.

I'm using Buster full.

**Comment by natinusala** [comment]
This looks like a bug in the core updater, I would fill another issue for that.

> How do I format my post so that I can present the commands I've tried the same way everyone else seems to?

Search for "github markdown" you will see how to do all the fancy formatting available on GitHub

> What is fkms-v3d?

It's the name of the device tree overlay to enable kms on the RPi3 and 4 (basically draw stuff to the screen without using X11)

> What is x11?

It's the display server on which the desktop manager relies on - it's an additional layer between RetroArch and the screen

> If I want to change the CLFAGS portion of my install, how should I go about recompiling?

Just make clean and run configure again

> Has anyone figured out a berryboot type method of dual booting the rpi4?

Google will have the answer to that

> I'd like to try RetroPie and see if I have the same issue.

Sadly RetroPie isn't available on RPi4 just yet, I believe Lakka is the only retrogaming distribution with a working port at the moment

**Comment by seven-lynx** [comment]
Thanks! That was a concise and clear response.

A small point, but I hate when someone says "Google it".

I did, and and know that the berryboot team is actually a long way from rpi4 support. I was just curious if someone had found a workaround, and if it would be simpler than trying to find whatever is going wrong with my current efforts.

Anyway, you answered all my minor questions, but I still don't know why retroarch isn't updating the cores or recognizing manually place core files. Obviously, it's possible to run retroarch on a rpi4 in Buster, because I've seen emu performance comparisons on this same setup. 

The best answer I've gotten so far is "Yes, the configuration you used worked fine for me."

Basically, what I'm, asking is, is there an error in the commands I used? Is there another step anyone took to get the rpi4 running retroarch?

The rpi4 is new enough that I'm not finding anyone with my exact error.

If you have gotten retroarch running emus on the rpi4, kindly tell me exactly what commands you used so I can figure out what's going wrong. I've even reformatted the whole sd card and started over, more than once.

I've tried every possible version of the methods details here and https://gist.github.com/AlexMax/32e5d038a66ce57253e740ea75736805

I'm starting to think there is a hardware fault, because I'm not even a good answer from the logs.

**Comment by natinusala** [comment]
How you compile retroarch does not affect the core updater unless you compile with --disable-ssl or any other network related flag.

May I suggest that you try Lakka? The RPi4 port is available as experimental images, you can find them in the latest blog post on lakka.tv

**Comment by zoltanvb** [comment]
It is an old issue, but still open. I'd assume compile problems are not around any more (see https://docs.libretro.com/guides/rpi/ for a current guide).

**Comment by gouchi** [comment]
@namanix Yes indeed, please close this issue. Thank you.
